src/lib/Microsoft.Health.Extensions.Fhir.R4/Service/ResourceManagementService.cs

```diff
@@ -69,13 +70,33 @@ protected async Task<TResource> CreateResourceByIdentityAsync<TResource>(Model.I
 
             propertySetter?.Invoke(resource, identifier);
 
-            return await FhirService.CreateResourceAsync(resource).ConfigureAwait(false);
+            // Generate id programatically and issue an update to FHIR service
+            // The resource will be created if the resource with the given id does not already exist.
+            resource.Id = ToBase64Variant(identifier);
+            return await FhirService.UpdateResourceAsync<TResource>(resource).ConfigureAwait(false);
         }
 
         private static Model.Identifier BuildIdentifier(string value, string system)
         {
             var identifier = new Model.Identifier { Value = value, System = string.IsNullOrWhiteSpace(system) ? null : system };
             return identifier;
         }
+
+        private static string ToBase64Variant(Model.Identifier identifier)
+        {
+            EnsureArg.IsNotNullOrWhiteSpace(identifier.Value, nameof(identifier.Value));
+
+            string plainTextId = $"{identifier.System}_{identifier.Value}";
+            var encodedId = Encoding.UTF8.GetBytes(plainTextId);
+            var base64 = Convert.ToBase64String(encodedId);
+
+            StringBuilder sb = new StringBuilder(base64);
+
+            sb.Replace("+", "-");
+            sb.Replace("/", "_");
+            sb.Replace("=", ".");
+
+            return sb.ToString().ToLower();

```

Comment:
Thanks Dustin - that is a very good catch regarding the valid fhir id characters and length, I had not thought about that. I think we should go with a sha256 hash. It will always result in a 64 character string with the characters [a-fA-F0-9] and I believe has the one of lowest chances of a collision. It seems to perform decently as well.